debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java

```diff
@@ -447,4 +458,34 @@ private Lsn getToLsn(SqlServerConnection connection, String databaseName, TxLogP
 
         return connection.getNthTransactionLsnFromLast(databaseName, fromLsn, maxTransactionsPerIteration);
     }
+
+    @Override
+    public void commitOffset(Map<String, ?> offset) {
+        Lsn commitLsn = Lsn.valueOf((String) offset.get("commit_lsn"));
+        synchronized (streamingExecutionContexts) {
+            for (Map.Entry<SqlServerPartition, SqlServerStreamingExecutionContext> entry : streamingExecutionContexts.entrySet()) {
+                SqlServerPartition partition = entry.getKey();
+                List<SqlServerChangeTable> changeTablesWithKnownStopLsn = entry.getValue().getChangeTablesWithKnownStopLsn();
+
+                synchronized (changeTablesWithKnownStopLsn) {
+                    List<SqlServerChangeTable> changeTablesToBeDeleted = changeTablesWithKnownStopLsn.stream()
+                            .filter(t -> t.getStopLsn().compareTo(commitLsn) < 0)
+                            .collect(Collectors.toList());
```

Comment:
The intermediate list is probably unnecessary. The actual deletion can happen in a `.forEach(t -> {})`.

Reply:
It throws a `java.util.ConcurrentModificationException` exception on an attempt to remove an element from within `forEach`.